### PR TITLE
1706 - Convert Clean Workplace to Polars 7 (convert create_purged_dfs_for_reconciliation_and_data function to Polars)

### PR DESCRIPTION
**Pull request by AchyutRatkanthiwar**
## Description
Trello ticket [#1706](https://trello.com/c/syxDFj9w/1706-clean-workplace7-write-polars-code-for-createpurgeddfsforreconciliationanddata-and-call-in-job)

Converted `create_purged_dfs_for_reconciliation_and_data` function to Polars and updated tests for the same.

## Testing
- [x] Unit tests passing
- [x] Successful [Job / Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1706-cln-wkp-7-Transform-ASCWDS-Data:1f571615-33c4-4490-9783-7d544aa4810d)
- [ ] Outputs checked in Athena
- [ ] Outputs compared using [SageMaker script](https://eu-west-2.console.aws.amazon.com/sagemaker/home?region=eu-west-2#/notebook-instances/data-engineering-notebook)
      - Notebook name: compare_outputs.ipynb
- [ ] Output Screenshots added to Trello ticket

## Migration to polars checklist
- [x] Check that similar utils functions don't already exist, or if one can be created
- [x] Check that utils are being saved in the appropriate place
- [x] Used LazyFrame option for test data
- [x] Unit tests added
- [x] Docstrings added
- [x] Added comment to pyspark functions which have been migrated with the link to the new location
      `# converted to polars -> filepath.py`
- [x] Updated CHANGELOG
- [x] Code reviewed by AI
- [x] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: projects/_01_ingest/ascwds/fargate/utils/clean_workplace_utils.py [hunk 1]
```diff
@@ -75,3 +77,102 @@ def valid_workplace_filter() -> pl.Expr:
         # exclude duplicate establishments
         ~pl.col(AWPClean.establishment_id).is_in(DUPLICATE_ESTABLISHMENT_IDS)
     )
+
+
+class CleanWorkplaceDataExpressions:
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/fargate/utils/clean_workplace_utils.py]: I think make this specific to purging function. 

In the job role estimates stuff, there's clean_utils which has FilterJobRoleGroupExpressions. So it's specific to the function using the class.

And then the doc string also make it more specific to purging.

File: projects/_01_ingest/ascwds/fargate/utils/clean_workplace_utils.py [hunk 2]
```diff
@@ -75,3 +77,102 @@ def valid_workplace_filter() -> pl.Expr:
...
+    """
+    This process is designed to purge/remove data which has not been updated for
+    an extended period of time.
+
+    If the worplace is a parent account, the mupddate used to purge is the
+    maximum of any account within that organisation.
+
+    The purge rules for workplace_last_active_date also takes last_logged_in
+    date into account.
+
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/fargate/utils/clean_workplace_utils.py:176]: Remove rows which have not been updated/logged into since the purge date. 

Rows in ASC-WDS workplace data are removed using master_update_date.
Rows in reconciliation data are removed using master_update_date and  last_logged_in_date (whichever is more recent).
Parent accounts use the most recent date from all accounts in that organisation.

File: projects/_01_ingest/unittest_data/polars_ingest_test_file_data.py [hunk 3]
```diff
@@ -1534,3 +1535,168 @@ class ValidateCqcLocations4FullLatestSnapshotTest:
         (RegistrationStatus.registered,),
         (date(2024, 1, 1),),
     ]
+
+
+@dataclass
+class CleanWorkplaceUtilsTestCase:
+    id: str
+    test_data: list[Any]
+    expected_data: list[Any]
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/unittest_data/polars_ingest_test_file_data.py:1543]: You can reduce the amount of test data by making test_data optional. If you look at polars_ind_cqc_test_file_data line 3010, that class is used in test cases, but not all of the items in the class are used all the time.

Then just drop the last column in each test.

**Review by RoyPriceCS50 — COMMENTED** [review]
Left some comments about reducing test data a bit.

File: projects/_01_ingest/unittest_data/polars_ingest_test_file_data.py [hunk 4]
```diff
@@ -1534,3 +1535,168 @@ class ValidateCqcLocations4FullLatestSnapshotTest:
...
+                ("org1", date(2024, 1, 1), date(2024, 1, 1), "No", date(2024, 1, 1)),
+                ("org1", date(2024, 6, 1), date(2024, 5, 1), "No", date(2024, 1, 1)),
+            ],
+            expected_data=[
+                ("org1", date(2024, 1, 1), date(2024, 1, 1), "No", date(2024, 1, 1), date(2024, 1, 1)),
+                ("org1", date(2024, 6, 1), date(2024, 5, 1), "No", date(2024, 1, 1), date(2024, 5, 1)),
+            ], # fmt: skip
+        ),
+        CleanWorkplaceUtilsTestCase(
+            id="single_record_org_uses_own_date",
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/unittest_data/polars_ingest_test_file_data.py]: This is the same case as org_max_is_scoped_per_import_date.

Show how it handles null master_update_date instead?

File: projects/_01_ingest/unittest_data/polars_ingest_test_file_data.py (hunk 3 above)
Comment (MarionJHolloway) [reply projects/_01_ingest/unittest_data/polars_ingest_test_file_data.py:1543]: There was no need for any test data, so I've removed that parameter completely.

File: projects/_01_ingest/ascwds/fargate/utils/clean_workplace_utils.py (hunk 1 above)
Comment (MarionJHolloway) [reply projects/_01_ingest/ascwds/fargate/utils/clean_workplace_utils.py]: Done, also updated the doc string

File: projects/_01_ingest/ascwds/tests/fargate/utils/test_clean_workplace_utils.py [hunk 5]
```diff
@@ -35,3 +44,125 @@ def test_valid_workplace_filter_returns_correct_rows(self):
         returned_lf = input_lf.filter(job.valid_workplace_filter())
 
         pl_testing.assert_frame_equal(expected_lf, returned_lf)
+
+
+class TestCleanWorkplaceDataExpressions:
+    exprs = job.CleanWorkplaceDataExpressions()
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/tests/fargate/utils/test_clean_workplace_utils.py]: TestPurgeWorkplaceDataExpressions

job.PurgeWorkplaceDataExpressions

File: projects/_01_ingest/ascwds/tests/fargate/utils/__pycache__/test_clean_workplace_utils.cpython-311-pytest-9.0.2.pyc.23524
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/tests/fargate/utils/__pycache__/test_clean_workplace_utils.cpython-311-pytest-9.0.2.pyc.23524:1]: Remove this ?

**Review by RoyPriceCS50 — COMMENTED** [review]
You've got merge conflict from patch ordering, and tests failed but just wants class name updating to new name. 

There's also a cache file sneaked in.

**Review by RoyPriceCS50 — APPROVED** [review]
Looks good.
